daemon/src/common/Profiler.cpp

```diff
@@ -0,0 +1,201 @@
+         * contains the name of the function, the type of sample (function start,
+         * function end, and if there's a new frame), and a timestamp.
+         */
+    std::vector <Profiler::Point>                   samples;
+    thread_local std::vector <Profiler::Point>      localSamples;
+    std::mutex mtx;
+
+    Sys::SteadyClock::time_point                    startTime;
+    bool                                            enabled    = false; ///TODO those bools are silly
+    bool                                            nextFrame  = false;
+    bool                                            shouldStop = false;
+
+
+
+    /// Called at the end of every frame
+    void Frame(){
+
+        if(nextFrame){
+            enabled=true;
+            shouldStop=true;
+            nextFrame=false;
+            startTime   = Sys::SteadyClock::now();
+            return;
+        }
+
+        if(shouldStop){
+            enabled=false;
+            shouldStop=false;
+            Stop();
+            return;
+        }
+
+        if(!enabled)
+            return;
+
+        std::lock_guard<std::mutex> lck (mtx);
+        samples.push_back(Point(FRAME,""));
+    }
+
+
+    /// Point saves the time it's constructed
+    Point::Point(Type type, const char *label=""): type(type), label(label){
+        auto timeElapsed = Sys::SteadyClock::now() - startTime;
+        time = std::chrono::duration_cast < std::chrono::microseconds > (timeElapsed).count();
+        tid = std::this_thread::get_id();
+    }
+
+
+    /// At the Beginning and End of every function
+
+    ProfilerGuard::ProfilerGuard( const char * label ): label( label ){
+        if(!enabled)
+            return;
+        localSamples.push_back( Point( START, label ) );
+    }
+
+    ProfilerGuard::~ProfilerGuard(){
+        if(!enabled)
+            return;
+        localSamples.push_back( Point( END, label ) );
+    }
+
+
+    void Stop(){
+        enabled=false;
+
+        auto file = FS::HomePath::OpenWrite("profile.log"); ///TODO add timestamp to file
+
+        std::sort(samples.begin(), samples.end());
+        std::string line;
+        std::string type;
+
+        for (auto& i : samples) {
+
+            switch(i.type){
+            case START:
+                type="B";
+                break;
+            case END:
+                type="E";
+                break;
+            case FRAME:
+                type="F";
+                break;
+            }
+
+            std::stringstream ss;
+            ss << std::to_string(i.time) << ";" << type << ";" << i.label << ";" << i.tid << "$\n";
+           // std::cout << ss.str();
+            file.Write(ss.str().c_str(), ss.str().length());
+        }
+
+        file.Close();
+        samples.clear();
+    }
+
+    void Sync(){ //TODO: optimize?
+        std::lock_guard<std::mutex> lck (mtx);
+
+        samples.insert(samples.end(), localSamples.begin(), localSamples.end());
```

Comment:
Would be nice to:
 - avoid allocations most frame
 - have the thread id so the visualizer can show multiple flame graphs
 - avoid copies when giving the samples

Here's a quick idea:
```cpp
size_t nextCapacity = local
Samples.capacity - std::max(localSamples.capacity() / 256, 1);
threadSamples.push_back(std::make_pair(threadID, std::move(localSamples)));
// At that point localSamples is empty
localSamples.reserve(nextCapacity);```